alibi/prototypes/protoselect.py

```diff
@@ -0,0 +1,543 @@
+import os
+import logging
+import numpy as np
+import matplotlib.pyplot as plt
+from matplotlib.offsetbox import OffsetImage, AnnotationBbox
+
+from tqdm import tqdm
+from copy import deepcopy
+from typing import Callable, Optional, Dict, List, Union, Tuple
+from sklearn.model_selection import KFold
+from sklearn.neighbors import KNeighborsClassifier
+from skimage.transform import resize
+
+from alibi.utils.distance import batch_compute_kernel_matrix
+from alibi.api.interfaces import Explainer, Explanation, FitMixin
+from alibi.api.defaults import DEFAULT_META_PROTOSELECT, DEFAULT_DATA_PROTOSELECT
+from alibi.utils.kernel import EuclideanDistance
+
+logger = logging.getLogger(__name__)
+
+
+class ProtoSelect(Explainer, FitMixin):
+    def __init__(self,
+                 eps: float,
+                 kernel_distance: Callable[[np.ndarray, np.ndarray], np.ndarray],
+                 lbd: Optional[float] = None,
+                 batch_size: int = int(1e10),
+                 preprocess_fn: Optional[Callable[[np.ndarray], np.ndarray]] = None,
+                 verbose: bool = False,
+                 **kwargs):
+        """
+        Constructor
+
+        Parameters
+        ----------
+        kernel_distance
+            Kernel distance to be used. Expected to support computation in batches.
+        eps
+            Epsilon ball size.
+        lbd
+            Penalty for each prototype. Encourages a lower number of prototypes to be selected.
+        batch_size
+            Batch size to be used for kernel matrix computation.
+        preprocess_fn
+            Preprocessing function for kernel matrix computation.
+        verbose
+             Whether to display progression bar while computing prototypes points.
+        """
+        super().__init__(meta=deepcopy(DEFAULT_META_PROTOSELECT))
+        self.kernel_distance = kernel_distance
+        self.eps = eps
+        self.lbd = lbd
+        self.batch_size = batch_size
+        self.preprocess_fn = preprocess_fn
+        self.verbose = verbose
+
+        # get kernel tag
+        if hasattr(self.kernel_distance, '__name__'):
+            kernel_distance_tag = self.kernel_distance.__name__
+        elif hasattr(self.kernel_distance, '__class__'):
+            kernel_distance_tag = self.kernel_distance.__class__.__name__
+        else:
+            kernel_distance_tag = 'unknown kernel distance'
+
+        # update metadata
+        self.meta['params'].update({
+            'kernel_distance': kernel_distance_tag,
+            'eps': eps,
+            'lbd': lbd,
+            'batch_size': batch_size,
+            'verbose': verbose
+        })
+
+    def fit(self,
+            X: np.ndarray,
+            X_labels: Optional[np.ndarray] = None,
+            Y: Optional[np.ndarray] = None) -> 'ProtoSelect':
```

Comment:
Let's discuss parameter naming here as usually `y` or `Y` plays the role of labels for `X`.

Reply:
Perhaps we could go with (as used in other functions) with:
`X` -> `X_ref`
`X_labels` -> `y_ref`
`Y` -> `X_test` or `X_proto` or `X`

Alternatively, if we think by far the most common use case would be to just have one dataset, we cold do:
`X` -> `X`
`X_labels` -> `y`
`Y` -> `X_test` or `X_proto`

Also compare with `alibi-detect` where we have `x_ref` and `x`.

Reply:
Decided to follow with the first suggestion as it is consistent with `alibi-detect`. For consistency reasons (see notation for counterfactual) I decided to change `y_ref` to `Y_ref`.